File: Books/A_Comprehensive_Approach_to_ML_and_AI/python/pca_simple_tutorial.py
import numpy as np
import matplotlib.pyplot as plt

np.random.seed(0)
N = 50
# Generate sample data
x_in = np.zeros((N, 2))
x_in[:, 0] = np.random.randn(N) + 1
x_in[:, 1] = 0.5 * np.random.randn(N) + 0.5

# Rotation matrix
theta = -0.707
R = np.array([[np.cos(theta), np.sin(theta)],
              [-np.sin(theta), np.cos(theta)]])
x_in = x_in @ R.T  # Rotate data

# Step 1: Calculate the mean and center the data
mu_x = np.mean(x_in, axis=0)
x_bar = x_in - mu_x

# Step 2: Calculate covariance matrix
C_x = np.cov(x_bar, rowvar=False)
print("Covariance of input:")
print(C_x)

# Step 3: Eigen-decomposition
eigvals, eigvecs = np.linalg.eig(C_x)
# Sort eigenvalues (and corresponding eigenvectors)
sorted_idx = np.argsort(eigvals)
d1, d2 = eigvals[sorted_idx[0]], eigvals[sorted_idx[1]]
e1 = eigvecs[:, sorted_idx[0]]
e2 = eigvecs[:, sorted_idx[1]]
print("Eigenvector e1:")
print(e1)
print("Eigenvector e2:")
print(e2)
print("Eigenvalue d1:")
print(d1)
print("Eigenvalue d2:")
print(d2)

# Plot centered input vectors
plt.figure()
plt.plot(x_in[:,0], x_in[:,1], 'o')
plt.plot(mu_x[0], mu_x[1], 'r+', markersize=12, markeredgewidth=2)
plt.title("Centered Input Vectors")
plt.axis("equal")
plt.xlim([-3,3]); plt.ylim([-3,3])
plt.show()

# Plot centered data with eigenvectors
plt.figure()
plt.plot(x_bar[:,0], x_bar[:,1], 'bo')
origin = np.array([[0, 0],[0, 0]])
plt.quiver(0, 0, 2*d1*e1[0], 2*d1*e1[1], color='r', angles='xy', scale_units='xy', scale=1, width=0.005)
plt.quiver(0, 0, 2*d2*e2[0], 2*d2*e2[1], color='r', angles='xy', scale_units='xy', scale=1, width=0.005)
plt.title("Centered Input Vectors with Eigenvectors")
plt.axis("equal")
plt.xlim([-3,3]); plt.ylim([-3,3])
plt.show()

# Step 4: Project input data onto principal components
# Compute projections: note that eigenvectors form the new basis.
y = np.vstack([e2, e1]) @ x_bar.T  # shape (2, N)

plt.figure()
plt.plot(y[0, :], y[1, :], 'o')
plt.title("Projections onto Principal Components")
plt.axis("equal")
plt.xlim([-3,3]); plt.ylim([-3,3])
plt.show()

# Step 5: Project data using only one principal component
y1 = e1.T @ x_bar.T  # projection onto first principal component
plt.figure()
plt.plot(y1, np.zeros_like(y1), 'o')
plt.title("Projections onto One Principal Component (e1)")
plt.axis("equal")
plt.xlim([-3,3]); plt.ylim([-3,3])
plt.show()

y2 = e2.T @ x_bar.T  # projection onto second principal component
plt.figure()
plt.plot(y2, np.zeros_like(y2), 'o')
plt.title("Projections onto One Principal Component (e2)")
plt.axis("equal")
plt.xlim([-3,3]); plt.ylim([-3,3])
plt.show()

File: Books/A_Comprehensive_Approach_to_ML_and_AI/R/pca_simple_tutorial.R
# PCA Simple Tutorial in R
#
# This demo uses Principal Component Analysis (PCA) to compute the principal components of the data.
# The data is first centered, the covariance matrix is computed, eigen decomposition is performed,
# and the results are visualized by plotting the data with overlaid eigenvectors and by projecting onto the principal components.

# For reproducibility
set.seed(0)

# -----------------------------
# Generate sample data
# -----------------------------
N <- 50
x_in <- cbind(rnorm(N) + 1, 0.5 * rnorm(N) + 0.5)
theta <- -0.707
R_mat <- matrix(c(cos(theta), sin(theta), -sin(theta), cos(theta)), nrow = 2)
x_in <- x_in %*% t(R_mat)

# Step 1: Calculate mean of input vectors
mu_x <- colMeans(x_in)

# Step 2: Center the data
x_bar <- sweep(x_in, 2, mu_x)

# Step 3: Calculate covariance matrix
C_x <- cov(x_bar)
cat("Covariance of input:\n"); print(C_x)

# Step 4: Eigen decomposition of covariance matrix
eig <- eigen(C_x)
V <- eig$vectors
D <- eig$values
e1 <- V[,1]
e2 <- V[,2]
cat("Eigenvector e1:\n"); print(e1)
cat("Eigenvector e2:\n"); print(e2)
cat("Eigenvalue d1:\n"); print(D[1])
cat("Eigenvalue d2:\n"); print(D[2])

# Plot centered data with mean marked
plot(x_in[,1], x_in[,2], pch = 16, main = "Centered Input Vectors", xlab = "X1", ylab = "X2", xlim = c(-3,3), ylim = c(-3,3))
points(mu_x[1], mu_x[2], col = "red", pch = 4, lwd = 2)

# Plot centered data with eigenvectors
plot(x_bar[,1], x_bar[,2], pch = 16, col = "blue", main = "Centered Data with Eigenvectors", xlab = "X1", ylab = "X2", xlim = c(-3,3), ylim = c(-3,3))
arrows(0, 0, 2 * sqrt(D[1]) * e1[1], 2 * sqrt(D[1]) * e1[2], col = "red", lwd = 2)
arrows(0, 0, 2 * sqrt(D[2]) * e2[1], 2 * sqrt(D[2]) * e2[2], col = "red", lwd = 2)

# Project data onto principal components
y_proj <- t(V) %*% t(x_bar)
plot(y_proj[1, ], y_proj[2, ], pch = 16, main = "Projection onto Principal Components",
     xlab = "PC1", ylab = "PC2", xlim = c(-3,3), ylim = c(-3,3))

# Project using only one principal component (e.g., PC1)
y_pc1 <- y_proj[1, ]
plot(y_pc1, rep(0, length(y_pc1)), pch = 16, main = "Projection onto PC1", xlab = "PC1", ylab = "", xlim = c(-3,3))
